### PR TITLE
Encapsulated gradient transform logic into separate class

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
Moved the logic for gradient coordinate transformations out of the GradientTransformComponent and Util files into a new GradientTransform class.  In the process, some of the transformation logic got condensed and simplified, and I was able to add a suite of unit tests to verify the transformation calculations directly.

By itself, the logic changes are a small optimization, but the real value from this change will come in a follow-up PR when the gradient classes are switched to use copies of a GradientTransform instance directly for coordinate transformations, instead of going through an EBus for each transformation, which can add up to millions of calls for terrain refreshes.

### CODE REVIEW

File: Gems/GradientSignal/Code/Include/GradientSignal/Ebuses/GradientTransformModifierRequestBus.h [hunk 1]
```diff
@@ -13,6 +13,25 @@
...
+        // world space but its origin will frozen at (3,3) in world space, no matter how much the parent moves around.
+        Local_ThisEntity,
+        // The gradient's origin is the world position of the reference entity.
+        World_ReferenceEntity,
+        // The gradient's origin is the local position of the reference entity, but in world space.
+        Local_ReferenceEntity,
+        // The gradient's origin is at (0,0,0) in world space.
+        World_Origin,
+        // The gradient's origin is in translated world space relative to the reference entity.
+        Relative,
```
Comment (NicholasVanSickle) [inline Gems/GradientSignal/Code/Include/GradientSignal/Ebuses/GradientTransformModifierRequestBus.h:32]: These can be `//!` to get picked up by Doxygen.

File: Gems/GradientSignal/Code/Include/GradientSignal/GradientTransform.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+namespace GradientSignal
+{
+    // This controls how the gradient repeats itself when queried outside the bounds of the shape.
+    enum class WrappingType : AZ::u8
+    {
+        None = 0,           // Unbounded - the gradient ignores the shape bounds.
+        ClampToEdge,        // The values on the edge of the shape will be extended outward in each direction.
+        Mirror,             // The gradient signal will be repeated but mirrored on every repeat.
+        Repeat,             // The gradient signal will be repeated in every direction.
+        ClampToZero,        // The value will always be 0 outside of the shape.
```
Comment (NicholasVanSickle) [inline Gems/GradientSignal/Code/Include/GradientSignal/GradientTransform.h:25]: Repeated: `//!` (or `///`) for Doxygen

**Review by NicholasVanSickle — APPROVED** [review]
Looks good to me; seems sensible to move away from bus overhead for this. Out of curiosity, have we tried any of the usual cache the handler / do a bus request with a lambda to perform multiple queries tricks for the gradient queries though?

**Comment by mbalfour-amzn** [comment]
> Looks good to me; seems sensible to move away from bus overhead for this. Out of curiosity, have we tried any of the usual cache the handler / do a bus request with a lambda to perform multiple queries tricks for the gradient queries though?

Quick answer: no.
Long answer: Caching the handler isn't really a great solution here.  The GradientTransformComponent lifetimes can change separately from the Gradient lifetimes, and the consumption of the transform happens on multiple threads.  So caching and using the handler requires mutexes, which is effectively the same overhead as the EBus.  The lifetime synchronization is also a bit touchier with handler caching, because using the handler outside its lifetime results in crashes, where using a stale GradientTransform outside the handler's lifetime just results in stale data.

I'll still unfortunately need a mutex around the cached GradientTransform instance, but I can at least use a shared_mutex, and the compiler has a bit more optimization information to work with as it knows exactly what function is being called to perform the transformation at compile time (i.e. GradientTransform::TransformPosition*), instead of always requiring a virtual call.  

With *either* solution, the bigger problem around it is that the gradient components themselves have a single GetValue() API, so the mutex overhead itself is taken on every call.  A second GetValues() API will be added soon that allows for a set of values to be queried in a single call, so the mutex overhead will be much smaller relative to the amount of work performed per call.

**Review by cgalvan — APPROVED** [review]

**Comment by NicholasVanSickle** [comment]
> > Looks good to me; seems sensible to move away from bus overhead for this. Out of curiosity, have we tried any of the usual cache the handler / do a bus request with a lambda to perform multiple queries tricks for the gradient queries though?
> 
> Quick answer: no. Long answer: Caching the handler isn't really a great solution here. The GradientTransformComponent lifetimes can change separately from the Gradient lifetimes, and the consumption of the transform happens on multiple threads. So caching and using the handler requires mutexes, which is effectively the same overhead as the EBus. The lifetime synchronization is also a bit touchier with handler caching, because using the handler outside its lifetime results in crashes, where using a stale GradientTransform outside the handler's lifetime just results in stale data.
> 
> I'll still unfortunately need a mutex around the cached GradientTransform instance, but I can at least use a shared_mutex, and the compiler has a bit more optimization information to work with as it knows exactly what function is being called to perform the transformation at compile time (i.e. GradientTransform::TransformPosition*), instead of always requiring a virtual call.
> 
> With _either_ solution, the bigger problem around it is that the gradient components themselves have a single GetValue() API, so the mutex overhead itself is taken on every call. A second GetValues() API will be added soon that allows for a set of values to be queried in a single call, so the mutex overhead will be much smaller relative to the amount of work performed per call.

Thanks for the explanation! Makes sense to add a `GetValues` API in general (nudges consumers towards doing the right thing, opens the door for more optimizations), but it's worth pointing out that you don't need to persistently cache the handler - you can do a lambda call to the bus and then receive a short-lived handler you can run multiple operations on:
```
GradientRequestBus::Event(
                    m_gradientId,
                    [](GradientRequestBus::Events* gradientRequests)
                    {
                        /*
                        * We now have the event handler instance and can make repeated calls without reacquiring the bus mutex
                        for (...)
                        {
                            gradientRequests->GetValue(...);
                        }
                        */
                    });
```
It's an occasionally handy trick if you do run into an EBus API not on the slate for refactoring but need to make a large volume of requests.

**Pull request merged**